core/lib/rom/relation/class_interface.rb

```diff
@@ -94,23 +94,35 @@ def schema(dataset = nil, as: nil, infer: false, &block)
           ds_name = dataset || schema_opts.fetch(:dataset, default_name.dataset)
           relation = as || schema_opts.fetch(:relation, ds_name)
 
-          name = Name[relation, ds_name]
+          @relation_name = Name[relation, ds_name]
           inferrer = infer ? schema_inferrer : nil
 
           unless schema_class
             raise MissingSchemaClassError.new(self)
           end
 
           dsl = schema_dsl.new(
-            name,
+            relation_name,
             schema_class: schema_class, attr_class: schema_attr_class, inferrer: inferrer,
             &block
           )
 
-          @schema = dsl.call
+          @schema_proc = dsl.method(:call).to_proc
+        elsif dataset.is_a?(Schema)
+          @schema = dataset
```

Comment:
How come `dataset` can be a schema instance? Looks like a workaround of some problem?